File: backend/app/services/component_geometry.py
from typing import Any, Dict, List


def scale_points_to_image_space(
    points: List[List[float]],
    image_width: int,
    image_height: int,
) -> List[List[float]]:
    """Map model output coords (0-1 or 0-1000) into image pixel space."""
    if not points:
        return points

    max_x = max(p[0] for p in points)
    max_y = max(p[1] for p in points)

    if max_x <= 1.0 and max_y <= 1.0 and (image_width > 1 or image_height > 1):
        return [[p[0] * image_width, p[1] * image_height] for p in points]

    if (
        max_x <= 1000
        and max_y <= 1000
        and (image_width > 1000 or image_height > 1000)
        and (max_x < image_width * 0.5 or max_y < image_height * 0.5)
    ):
        return [
            [(p[0] / 1000.0) * image_width, (p[1] / 1000.0) * image_height]
            for p in points
        ]

    if max_x > image_width * 1.05 or max_y > image_height * 1.05:
        if max_x <= 1000 and max_y <= 1000:
            return [
                [(p[0] / 1000.0) * image_width, (p[1] / 1000.0) * image_height]
                for p in points
            ]
        scale_x = image_width / max_x if max_x else 1.0
        scale_y = image_height / max_y if max_y else 1.0
        scale = min(scale_x, scale_y)
        return [[p[0] * scale, p[1] * scale] for p in points]

    return points


def _polygon_area(polygon: List[List[float]]) -> float:
    if len(polygon) < 3:
        return 0.0
    area = 0.0
    n = len(polygon)
    for i in range(n):
        j = (i + 1) % n
        area += polygon[i][0] * polygon[j][1]
        area -= polygon[j][0] * polygon[i][1]
    return abs(area) / 2.0


def _polygon_bbox(polygon: List[List[float]]) -> List[float]:
    xs = [p[0] for p in polygon]
    ys = [p[1] for p in polygon]
    min_x, max_x = min(xs), max(xs)
    min_y, max_y = min(ys), max(ys)
    return [min_x, min_y, max_x - min_x, max_y - min_y]


def normalize_components_coordinates(
    components: List[Dict[str, Any]],
    image_width: int,
    image_height: int,
) -> List[Dict[str, Any]]:
    """Normalize all component polygons to image pixel space."""
    if not components or not image_width or not image_height:
        return components

    normalized: List[Dict[str, Any]] = []
    for comp in components:
        polygon = comp.get("polygon") or []
        scaled = scale_points_to_image_space(polygon, image_width, image_height)
        if scaled == polygon:
            normalized.append(comp)
            continue

        updated = dict(comp)
        updated["polygon"] = scaled
        updated["bbox"] = _polygon_bbox(scaled)
        updated["area_pixels"] = _polygon_area(scaled)
        normalized.append(updated)

    return normalized

File: frontend/src/utils/componentGeometry.ts
import type { Component } from '../services/api';

export function polygonArea(polygon: number[][]): number {
  const n = polygon.length;
  if (n < 3) return 0;
  let sum = 0;
  for (let i = 0; i < n; i++) {
    const [x1, y1] = polygon[i];
    const [x2, y2] = polygon[(i + 1) % n];
    sum += x1 * y2 - x2 * y1;
  }
  return Math.abs(sum) / 2;
}

export function polygonBbox(polygon: number[][]): number[] {
  if (polygon.length === 0) return [0, 0, 0, 0];
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  for (const [x, y] of polygon) {
    if (x < minX) minX = x;
    if (y < minY) minY = y;
    if (x > maxX) maxX = x;
    if (y > maxY) maxY = y;
  }
  return [minX, minY, maxX - minX, maxY - minY];
}

export function bboxToPolygon(bbox: number[]): number[][] {
  const [x, y, w, h] = bbox;
  return [
    [x, y],
    [x + w, y],
    [x + w, y + h],
    [x, y + h],
  ];
}

export function updateComponentBbox(comp: Component, bbox: number[]): Component {
  const [x, y, w, h] = bbox;
  const bboxArea = w * h;
  return {
    ...comp,
    bbox: [x, y, w, h],
    polygon: bboxToPolygon(bbox),
    area_pixels: comp.mask_base64 ? comp.area_pixels : bboxArea,
  };
}

export function updateComponentMask(
  comp: Component,
  mask_base64: string,
  area_pixels: number,
): Component {
  return {
    ...comp,
    mask_base64,
    area_pixels,
  };
}

export function clampBbox(
  bbox: number[],
  refWidth: number,
  refHeight: number,
  minSize = 20,
): number[] {
  let [x, y, w, h] = bbox;
  w = Math.max(minSize, w);
  h = Math.max(minSize, h);
  x = Math.max(0, Math.min(x, refWidth - w));
  y = Math.max(0, Math.min(y, refHeight - h));
  return [x, y, w, h];
}

export type ResizeHandle = 'nw' | 'ne' | 'sw' | 'se';

export function resizeBbox(
  startBbox: number[],
  handle: ResizeHandle,
  dx: number,
  dy: number,
  refWidth: number,
  refHeight: number,
): number[] {
  let [x, y, w, h] = startBbox;

  if (handle.includes('e')) w += dx;
  if (handle.includes('w')) {
    x += dx;
    w -= dx;
  }
  if (handle.includes('s')) h += dy;
  if (handle.includes('n')) {
    y += dy;
    h -= dy;
  }

  return clampBbox([x, y, w, h], refWidth, refHeight);
}

export function moveBbox(
  startBbox: number[],
  dx: number,
  dy: number,
  refWidth: number,
  refHeight: number,
): number[] {
  const [x, y, w, h] = startBbox;
  return clampBbox([x + dx, y + dy, w, h], refWidth, refHeight);
}
